src/shift.c

```diff
@@ -23,194 +23,146 @@ SEXP shift(SEXP obj, SEXP k, SEXP fill, SEXP type) {
             INTEGER(tmp)[m] = INTEGER(thisfill)[0];
-          copyMostAttrib(elem, tmp);
-          if (isFactor(elem))
-            setAttrib(tmp, R_LevelsSymbol, getAttrib(elem, R_LevelsSymbol));
-        }
-        break;
-
-      case REALSXP :
-        klass = getAttrib(elem, R_ClassSymbol);
-        if (isString(klass) && STRING_ELT(klass, 0) == char_integer64) {
-          thisfill = PROTECT(allocVector(REALSXP, 1)); protecti++;
-          dthisfill = (unsigned long long *)REAL(thisfill);
-          if (INTEGER(fill)[0] == NA_INTEGER)
-            dthisfill[0] = NA_INT64_LL;
-          else dthisfill[0] = (unsigned long long)INTEGER(fill)[0];
+          // only two possibilities left: type = 'lead', n>=0 _or_ type = 'lag', n<0
         } else {
-          thisfill = PROTECT(coerceVector(fill, REALSXP)); protecti++;
-        }
-        for (j=0; j<nk; j++) {
-          thisk = (xrows >= INTEGER(k)[j]) ? INTEGER(k)[j] : xrows;
-          SET_VECTOR_ELT(ans, i*nk+j, tmp=allocVector(REALSXP, xrows) );
-          if (xrows - INTEGER(k)[j] > 0) {
-            memmove((char *)DATAPTR(tmp)+(INTEGER(k)[j]*size),
-                 (char *)DATAPTR(elem),
-                 (xrows-INTEGER(k)[j])*size);
-          }
-          for (m=0; m<thisk; m++) {
-            REAL(tmp)[m] = REAL(thisfill)[0];
-          }
-          copyMostAttrib(elem, tmp);
-        }
-        break;
-
-      case LGLSXP :
-        thisfill = PROTECT(coerceVector(fill, LGLSXP)); protecti++;
-        for (j=0; j<nk; j++) {
-          thisk = (xrows >= INTEGER(k)[j]) ? INTEGER(k)[j] : xrows;
-          SET_VECTOR_ELT(ans, i*nk+j, tmp=allocVector(LGLSXP, xrows) );
-          if (xrows - INTEGER(k)[j] > 0)
-            memmove((char *)DATAPTR(tmp)+(INTEGER(k)[j]*size),
-                 (char *)DATAPTR(elem),
-                 (xrows-INTEGER(k)[j])*size);
-          for (m=0; m<thisk; m++)
-            LOGICAL(tmp)[m] = LOGICAL(thisfill)[0];
-          copyMostAttrib(elem, tmp);
-        }
-        break;
-
-      case STRSXP :
-        thisfill = PROTECT(coerceVector(fill, STRSXP)); protecti++;
-        for (j=0; j<nk; j++) {
-          SET_VECTOR_ELT(ans, i*nk+j, tmp=allocVector(STRSXP, xrows) );
-          for (m=0; m<xrows; m++)
-            SET_STRING_ELT(tmp, m, (m < INTEGER(k)[j]) ? STRING_ELT(thisfill, 0) : STRING_ELT(elem, m - INTEGER(k)[j]));
-          copyMostAttrib(elem, tmp);
-        }
-        break;
-
-      case VECSXP :
-        thisfill = PROTECT(coerceVector(fill, VECSXP)); protecti++;
-        for (j=0; j<nk; j++) {
-          SET_VECTOR_ELT(ans, i*nk+j, tmp=allocVector(VECSXP, xrows) );
-          for (m=0; m<xrows; m++)
-            SET_VECTOR_ELT(tmp, m, (m < INTEGER(k)[j]) ? VECTOR_ELT(thisfill, 0) : VECTOR_ELT(elem, m - INTEGER(k)[j]));
-          copyMostAttrib(elem, tmp);
-        }
-        break;
-
-      default :
-        error("Unsupported type '%s'", type2char(TYPEOF(elem)));
-      }
-      copyMostAttrib(elem, tmp);
```

Comment:
I honestly thing these lines are vestigial and I think they may be slowing down `shift` unnecessarily... there's no analogue in the `'lead'` branch and I think what it's accomplishing is already done in the `INTSXP` branch.

Reply:
which "these" precisely? what is it is accomplishing that is already done in INTSXP?